### PR TITLE
Bug-fixes value-defaulting

### PR DESCRIPTION
**Pull request by georgekuruvillak**
**What this PR does / why we need it**:
This PR adds default values to values.yaml. It also takes care of go-template value conversion issues. 
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Added default values to values.yaml. It also takes care of go-template value conversion issues.
```

### CODE REVIEW

**Comment by georgekuruvillak** [comment]
@amshuman-kr There were some bugs fixes associated with conversion of values that resulted in incorrent conversions example int conversion of size resources etc. PTAL. I have created a hotfix as well. Will trigger the release post to the review.

File: charts/etcd/templates/etcd-statefulset.yaml [hunk 1]
```diff
@@ -113,15 +113,23 @@ spec:
         command:
         - etcdbrctl
         - server
+{{- if .Values.etcd.defragmentationSchedule }}
         - --defragmentation-schedule={{ .Values.etcd.defragmentationSchedule }}
+{{- end }}
+{{- if .Values.backup.fullSnapshotSchedule }}
         - --schedule={{ .Values.backup.fullSnapshotSchedule }}
+{{- end }}
         - --garbage-collection-policy={{ .Values.backup.garbageCollectionPolicy }}
```
Comment (amshuman-kr) [inline charts/etcd/templates/etcd-statefulset.yaml]: Why not a check for this too?

File: charts/etcd/templates/etcd-statefulset.yaml [hunk 2]
```diff
@@ -132,22 +140,25 @@ spec:
...
         - --endpoints=https://{{ .Values.name }}-local:{{ .Values.etcd.clientPort }}
-        # enable TLS on backup-restore server reusing etcd cert bundle
-        - --server-cert=/var/etcd/ssl/server/tls.crt
-        - --server-key=/var/etcd/ssl/server/tls.key
 {{ else }}
         - --insecure-transport=true
         - --insecure-skip-tls-verify=true
         - --endpoints=http://{{ .Values.name }}-local:{{ .Values.etcd.clientPort }}
 {{- end }}
         - --etcd-connection-timeout={{ .Values.backup.etcdConnectionTimeout }}
```
Comment (amshuman-kr) [inline charts/etcd/templates/etcd-statefulset.yaml]: Why not a check for this too?

File: controllers/etcd_controller.go [hunk 3]
```diff
@@ -883,6 +889,27 @@ func (r *EtcdReconciler) removeFinalizersToDependantSecrets(etcd *druidv1alpha1.
 				}
 			}
 		}
+
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: Why copy-paste like this? Why not run this in a loop for all the secrets what need to be updated?

File: charts/etcd/templates/etcd-statefulset.yaml (hunk 1 above)
Comment (georgekuruvillak) [reply charts/etcd/templates/etcd-statefulset.yaml]: For `garbageCollectionPolicy`, we get a default value from values.yaml if its absent.

File: charts/etcd/templates/etcd-statefulset.yaml (hunk 2 above)
Comment (georgekuruvillak) [reply charts/etcd/templates/etcd-statefulset.yaml]: For `etcdConnectionTimeout`, we take a default value from values.yaml if absent.

File: controllers/etcd_controller.go [hunk 4]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
-				storeSecretCopy.Finalizers = finalizers.UnsortedList()
-				if err := r.Update(context.TODO(), storeSecretCopy); err != nil {
-					return err
-				}
-			}
-		}
-	}
 
-	// As the secrets inside TLS field are required, we error in case it is not found as well.
+	tlsSecrets := []string{}
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
	secrets := []string{}
```
The backup secret is not TLS secret 😉

File: controllers/etcd_controller.go [hunk 5]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
-			logger.Infof("Adding finalizer (%s) for secret %s by etcd (%s)", FinalizerName, serverSecret.GetName(), etcd.Name)
-			serverSecretCopy := serverSecret.DeepCopy()
-			finalizers.Insert(FinalizerName)
-			serverSecretCopy.Finalizers = finalizers.UnsortedList()
-			if err := r.Update(context.TODO(), serverSecretCopy); err != nil {
-				return err
-			}
-		}
+		// As the secrets inside TLS field are required, we error in case they are not found.
+		tlsSecrets = append(tlsSecrets,
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
		secrets = append(secrets,
```

File: controllers/etcd_controller.go [hunk 6]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
+			etcd.Spec.Etcd.TLS.ClientTLSSecretRef.Name,
+			etcd.Spec.Etcd.TLS.ServerTLSSecretRef.Name,
+			etcd.Spec.Etcd.TLS.TLSCASecretRef.Name,
+		)
+	}
+	if etcd.Spec.Backup.Store != nil && etcd.Spec.Backup.Store.SecretRef != nil {
+		// As the store secret is required, we error in case it is not found as well.
+		tlsSecrets = append(tlsSecrets, etcd.Spec.Backup.Store.SecretRef.Name)
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
		secrets = append(secrets, etcd.Spec.Backup.Store.SecretRef.Name)
```

File: controllers/etcd_controller.go [hunk 7]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
+	}
 
-		caSecret := corev1.Secret{}
+	for _, tlsSecret := range tlsSecrets {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
	for _, secretName := range secrets {
```

File: controllers/etcd_controller.go [hunk 8]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
+		secret := corev1.Secret{}
 		if err := r.Client.Get(context.TODO(), types.NamespacedName{
-			Name:      etcd.Spec.Etcd.TLS.TLSCASecretRef.Name,
-			Namespace: etcd.Spec.Etcd.TLS.TLSCASecretRef.Namespace,
-		}, &caSecret); err != nil {
+			Name:      tlsSecret,
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
			Name:      secretName,
```

File: controllers/etcd_controller.go [hunk 9]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
-				finalizers.Delete(FinalizerName)
-				storeSecretCopy.Finalizers = finalizers.UnsortedList()
-				if err := r.Update(context.TODO(), storeSecretCopy); err != nil && !errors.IsNotFound(err) {
-					return err
-				}
-			}
-		}
-	}
+
+	tlsSecrets := []string{}
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
	secrets := []string{}
```

File: controllers/etcd_controller.go [hunk 10]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
-				logger.Infof("Removing finalizer (%s) from secret %s", FinalizerName, clientSecret.GetName())
-				clientSecretCopy := clientSecret.DeepCopy()
-				finalizers.Delete(FinalizerName)
-				clientSecretCopy.Finalizers = finalizers.UnsortedList()
-				if err := r.Update(context.TODO(), clientSecretCopy); err != nil && !errors.IsNotFound(err) {
-					return err
-				}
-			}
-		}
+		tlsSecrets = append(tlsSecrets,
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
		secrets = append(secrets,
```

File: controllers/etcd_controller.go [hunk 11]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
+			etcd.Spec.Etcd.TLS.ClientTLSSecretRef.Name,
+			etcd.Spec.Etcd.TLS.ServerTLSSecretRef.Name,
+			etcd.Spec.Etcd.TLS.TLSCASecretRef.Name,
+		)
+	}
+	if etcd.Spec.Backup.Store != nil && etcd.Spec.Backup.Store.SecretRef != nil {
+		tlsSecrets = append(tlsSecrets, etcd.Spec.Backup.Store.SecretRef.Name)
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
		secrets = append(secrets, etcd.Spec.Backup.Store.SecretRef.Name)
```

File: controllers/etcd_controller.go [hunk 12]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
+	}
 
-		serverSecret := corev1.Secret{}
-		err = r.Client.Get(context.TODO(), types.NamespacedName{
-			Name:      etcd.Spec.Etcd.TLS.ServerTLSSecretRef.Name,
+	for _, tlsSecret := range tlsSecrets {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
	for _, secretName := range secrets {
```

File: controllers/etcd_controller.go [hunk 13]
```diff
@@ -740,150 +746,79 @@ func (r *EtcdReconciler) getMapFromEtcd(etcd *druidv1alpha1.Etcd) (map[string]in
...
+		secret := corev1.Secret{}
+		err := r.Client.Get(context.TODO(), types.NamespacedName{
+			Name:      tlsSecret,
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
			Name:      secretName,
```

**Comment by amshuman-kr** [comment]
@georgekuruvillak I don't see the variable name changes you marked as resolved.

**Pull request merged**